ocaml/parsing/parser.mly

```diff
@@ -822,6 +813,47 @@ let package_type_of_module_type pmty =
   | _ ->
       err pmty.pmty_loc Neither_identifier_nor_with_type
 
+(* There's no dedicated syntax for module instances. Functor application
+   syntax is translated into a module instance expression.
+*)
+let pmod_instance : module_expr -> module_expr_desc =
+  let raise_malformed_instance loc =
+    raise (Syntaxerr.Error (Malformed_instance_identifier loc))
+  in
+  let head_of_ident (lid : Longident.t Location.loc) =
+    match lid with
+    | { txt = Lident s; loc = _ } -> s
+    | { txt = _; loc } ->  raise_malformed_instance loc
+  in
+  let gather_args mexpr =
+    let rec loop mexpr acc =
+      match mexpr with
+      | { pmod_desc = Pmod_apply (f, v); _ } ->
+         (match f.pmod_desc with
+          | Pmod_apply (f, n) -> loop f ((n, v) :: acc)
+          | _ -> raise_malformed_instance f.pmod_loc)
+      | head -> head, acc
```

Comment:
FYI @lukemaurer : this loop used to reverse the list at the end, but I think this was unintentional: the list is being built up in the proper order already bc functor application left-associates. I've changed this in this PR. Probably it doesn't matter anyway, as the order of instance args doesn't appear to mean anything.

Reply:
!! Yes, I think you're right on all counts (including that it doesn't matter much, though for errors we probably do want to stick to the user's syntax?).